### DISCUSSION TITLE
Local Web Applications with Traefik reverse proxy

### DISCUSSION THREAD

**Discussion by amyb12345** [docker/extension-ideas#28] [Ideas]
- Reverse proxy and security
 - Easy setup
 
Note: This is a transfer from our initial method of collecting new extension requests. If you have additional information on how this will help you or other functionality you'd like to see, please add more details in the comments.

**Comment by ijpatricio** [comment]
Hey Amy. I've been a fan of Traefik and Docker Compose for a while now. I could give it a try to make such an extension.

Can you expand further on this idea??

The extension would enable a sensible default Traefik configuration so that local web applications would be accessible by a domain.
Apart from this, generating an SSL cert and authorizing it within the OS, locally?

Are these more or less the lines of the idea?

**Comment by amyb12345** [comment]
Hi! Many of the ideas that I added to this list were submitted by Docker users in a previous format we were using.  The format was a survey form in product that we realized didn't get us the opportunity to follow up with the user who submitted the idea or get more details on the use cases.  It did leave the ability for users to leave contact info and if they did, I did follow up with them and ask them to come here and put more detail. Unfortunately, it looks like they haven't yet. :(  With that being said, I went back through the results and found "to be able to easily setup reverse proxy and security" , "HTTP reverse proxy", "easy setup and configuration for Traefik" as the more descriptive responses over just "Traefik".  You can definitely see why we switched to this forum!  Here we can get in touch with those who upvote as well as ask more clarifying issues on the discussion. 

With all that background, and based on what you described above, I think  what you are proposing would be a good approach.  If we built that and included it in our Marketplace, we can always have a place in the extension for additional feedback or share the repo with them to suggest improvements, etc.

**Comment by ijpatricio** [comment]
Hey Amy,

Yeah, I agree that this topic should not be called "Traefik". Many would not understand the benefit. I'm thinking more along the lines of "Local Web Applications".

Goals/benefits of the Extension:

- Virtual hosts: Multiple web applications served on the default port 80.
- Enabling local certificates, so we can have SSL while developing. Good to mimic prod environments or tinker with SSL-only enabled browser APIs.
- Edit hosts: Add local domains that would then point to 127.0.0.1.
- Dashboard: Display what routes are registered and which containers are bonded.

What do you think of changing the topic to `Local Web Applications with Traefik reverse proxy` so that it is straightforward for everyone in the audience?

I want to be able to develop this as part of Youtube sessions in my channel and also most likely to build it with a backend and database, meaning it's the more advanced extension.

Does all of this sound good to you?

**Reply by ijpatricio** [reply]
Ups, posted in the middle of writing. Editing now...

**Reply by ijpatricio** [reply]
Ok, @amyb12345 just finished editing :)

**Reply by amyb12345** [reply]
@ijpatricio I love that idea. I've made the update to the title as you suggested. If you need any help or have questions as you are developing, please don't hesitate to reach out to our extensions team on Slack, email or the forums!  We are happy to help.  I think this could be high value to the community!

**Reply by ijpatricio** [reply]
@amyb12345 great! Very glad to hear! I'll start working on this next week and keep you updated here.

So... 😄 what's a good name for this extension?? lol

Happy new year! 🎉

**Reply by amyb12345** [reply]
@ijpatricio I spoke with a few developers inside of Docker to get some name suggestions.  They suggested "Traefik Reverse Proxy", but mentioned that we should work on a good  short and detailed descriptions to elaborate on the use cases the extension will solve.  This will allow those not familiar with Traefik to still find the extension based on their searches. 

Happy new year as well!!

**Reply by ijpatricio** [reply]
@amyb12345 that's precisely it! More on that, I think there's (hypothetically) a place for two extensions:

- number one is a (dedicated) Traefik extension to dive deep into Traefik's features
- number two is a tool to help with the local development of web applications. I'm considering using Traefik, but in the future, that might change for something else, or Traefik may become one of the multiple adapters: we'll add Nginx, Caddy, and HAproxy...

I'd like to collaborate with option number two. The "first party citizen" is the problem-solving, not the tooling.

I also suggest we keep it simple to be achievable. We can iterate later, if at all 😄

**Comment by amyb12345** [comment]
Sounds like a good plan! I like the idea of having multiple adapters in the future!

**Reply by ijpatricio** [reply]
Hey @amyb12345 

So yesterday, I gave it a go and built an extension :) 

Tinkered with the examples with React and Go. Then I gave it a shot with the stack where I feel more productive: Laravel and Vue or Livewire (photo below: development, and extension containers visibility = on).

[Undocumented repo here](https://github.com/finuras/sidecar-docker-extension)

So from here, I can quickly achieve the goal we've discussed. But I ran into an issue that might be a problem: the image size is ~800Mb-1Gb.

I can achieve this with React or Vue inside the Electron Node process and get a host Docker API; I've tinkered with it. I'm just afraid I'll be much, much slower.

So I wanted to check in with you, as it's ONLY for development, is this size a problem? There's also the possibility could be optimized, and I don't know how to do it.

![image](https://user-images.githubusercontent.com/26031459/210334193-81baeeac-3773-483c-9747-7097ebc7c641.png)

**Reply by ivan-curkovic** [reply]
Hi @ijpatricio, 
for laravel, can you try to use production docker file (instead of dev).
https://github.com/finuras/raft/blob/6c3155f14589f5b50dda030c957dc681ba4bc511/server/docker-compose.yml#L7

**Reply by ijpatricio** [reply]
Hey @ivan-curkovic 

Ok, I'll do just that! My concern would be that it would be preferable to develop the extension without using a container, just UI and Node process from the base app. (is it electron?)

As for me, I'm ok (perfect) with that, I prefer this way because I can use my everyday tools and be so much faster!

Thanks, I'll put my hands to work, speak soon @ivan-curkovic @amyb12345

**Reply by ijpatricio** [reply]
Hello @ivan-curkovic 

So I went ahead and optimized the image the most I could. The results:

![image](https://user-images.githubusercontent.com/26031459/211356722-61aed148-ccb5-4dd4-bef1-a3ba74fc4b96.png)


Here is the gist: (let me know if you want more details)

This project needs to run Docker commands on the host (for example, start the Traefik container).
Therefore, I'm mapping the host socket `/var/run/docker.sock` inside the extension container (`raft-server`).
Using an optimized base image (serversideup/php - 211MB) with the "raw" project (tag on the screenshot: `ssu-no-docker-cli`) I get a size of 264MB.
The thing is, to run docker commands, I'm installing `docker-ce-cli and docker-compose-plugin`, and this brings the image size up to 527MB.

`docker-ce-cli` installed size: ~69MB after  [reference](https://helpmanual.io/packages/apt/docker-ce-cli/)
`docker-compose-plugin` installed size 25MB [reference](https://ubuntu.pkgs.org/20.04/docker-ce-amd64/docker-compose-plugin_2.5.0~ubuntu-focal_amd64.deb.html)

I'm already happy with ~500MB because it's for local usage only.
Also, by using Laravel, I can develop the extension much faster.

But is it possible to improve it further? Another way not installing `docker-cli`?

**Reply by ijpatricio** [reply]
Hello @ivan-curkovic 
After speaking with @ajeetraina , I changed the extension name as suggested (not to be ambiguous with Raft Consensus).
The new repo is here https://github.com/finuras/sidecar-docker-extension

I've been making progress with the extension using this image with this size, and everything is great and smooth.
Of course, if there's a way to have docker-cli commands with smaller size, let me know, I'll change accordingly!

Thanks!

**Reply by ivan-curkovic** [reply]
Hi @ijpatricio, I am checking with the docker team if we have some ideas to make it smaller or remove it.

**Reply by ijpatricio** [reply]
Hi @ivan-curkovic ! Thank you very much for taking that time, that's awesome!

A side note, because of the Excalidraw extension, I can see you're a fan of it! I'm a huge fan. Also, I turn a lot of people into Excalidraw users eheheh even non-tech-savvy people, like clients, managers, and friends 😄 I'm a subscriber, so I use the cloud app, but I can recommend the extension to Docker users now 👍 that was genius! Thanks!!

**Comment by ijpatricio** [comment]
Hi @amyb12345 @ivan-curkovic 

I left the other thread to discuss ways to optimize further the image size.

Here I want to let you see in a quick way the current state of the thing.

I got out of time to record again, so I had to sneak it the last seconds all that I could. Pause the video to see the file in detail because there's just a few seconds of it 😄 

Here's the video (5min) and the repo. Let me know what you think! 

https://www.loom.com/share/87dd39cf7fe44bba9bfd8b1019c2ee1e

https://github.com/finuras/raft

Thanks

PS: I only cared a little about the UI. When it gets good feature-wise, we can make it resemble Docker-UI/MaterialUI.

**Reply by ijpatricio** [reply]
Also, the config file is published in the home folder.

![image](https://user-images.githubusercontent.com/26031459/211359148-a51e4fea-47e4-4c3e-a9d3-6f11c33e2b80.png)

**Reply by amyb12345** [reply]
@ijpatricio This looks great so far! I also shared this with one of our Dev. Rel. team members to take a look. Very nice progress!

**Reply by ijpatricio** [reply]
@amyb12345 thank you for the kind comment! 😄  That's fantastic news. I can't wait to hear further feedback! I'm getting a little busy until the end of the month, but I'll try to sneak in some time for this even so.

**Reply by ijpatricio** [reply]
Hey @amyb12345 . So @ajeetraina and I are already in contact and have made a few improvements!

The repository link has changed, it is now https://github.com/finuras/sidecar-docker-extension

It's time to improve the UX, make it intuitive, and make a User manual with examples. Then I can also make a demo on Youtube.

Do let me know anything more, thanks!!

**Comment by amyb12345** [comment]
Awesome! I think you've got it covered!

**Reply by ijpatricio** [reply]
This is a first time, so.... not so sure eheheh I hope the same, let's see 😄 

I'll keep you posted, please do the same if any needs arise 👍

**Reply by ajeetraina** [reply]
Thanks @ijpatricio for building this extension. Loved the great conversations that we had to shape up this extension.

**Reply by ijpatricio** [reply]
Hello @ajeetraina, no problem, mate! My pleasure. I think there's still work to do, right?

How can I now get beta testers? I can find people to test it and provide feedback, but should be available from marketplace, or a link.
Something easy, not cloning, running commands, etc

After this I could refine the workflow, and make a Instructions Website

what do you think about this?

**Reply by amyb12345** [reply]
@ijpatricio Once you install your extension, you should be able to generate a shareable URL in Docker Desktop.  If you go to Manage Extensions, you should see a Share button that will copy a URL to make it a one click install for users!  Let me know if this works.  I'd be happy to also share a message to my social networks to help you get beta testers and feedback if that's helpful too!

**Reply by ijpatricio** [reply]
Hey @amyb12345, Great, I didn't know about this! Woohoo! So I'm leaving tomorrow for a conference (LaraconEU), and I'll be back next week. I will explore this and try to find a few friends to experiment with.
Thank you so much; I find it very helpful, to have feedback from "unknown testers", to which I can't speak, or introduce the tool. No better test than that 😄 
I'll ping you back next week, ok? Thank you so much, have a great one!

**Reply by amyb12345** [reply]
Enjoy your trip and conference!
